### PR TITLE
Remove conditional and broadcast for signalling consumers

### PR DESCRIPTION
**Pull request by derekcollison**
Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

File: server/consumer.go [hunk 1]
```diff
@@ -1238,7 +1247,8 @@ func (o *Consumer) processNextMsgReq(_ *subscription, c *client, _, reply string
 	if o.replay {
 		o.waiting.add(&wr)
 		o.mu.Unlock()
-		mset.signalConsumers()
+		o.signalNewMessages()
+		//mset.signalConsumers()
```
Comment (kozlovic) [inline server/consumer.go:1251]: Remove?

File: server/consumer.go [hunk 2]
```diff
@@ -1238,7 +1247,8 @@ func (o *Consumer) processNextMsgReq(_ *subscription, c *client, _, reply string
 	if o.replay {
 		o.waiting.add(&wr)
 		o.mu.Unlock()
-		mset.signalConsumers()
+		o.signalNewMessages()
```
Comment (kozlovic) [inline server/consumer.go:1250]: In some places this is invoked inside of consumer lock, but here it is outside. Note that in signalNewMessages() you are accessing consumer's field that may or may not be immutable, not sure.

File: server/stream.go [hunk 3]
```diff
@@ -1013,25 +1009,12 @@ func (mset *Stream) processInboundJetStreamMsg(_ *subscription, pc *client, subj
 		for _, o := range obs {
 			o.incStreamPending(seq, subject)
 			if !o.deliverCurrentMsg(subject, hdr, msg, seq, ts) {
-				needSignal = true
+				o.signalNewMessages()
```
Comment (kozlovic) [inline server/stream.go:1012]: Should we have a locked/unlocked version?

**Review by kozlovic — COMMENTED** [review]
Not sure about locking...

File: server/consumer.go (hunk 1 above)
Comment (derekcollison) [reply server/consumer.go:1251]: geez, I even looked for all of them ;)

Comment (derekcollison) [reply server/consumer.go:1251]: Will fix.

File: server/consumer.go (hunk 2 above)
Comment (derekcollison) [reply server/consumer.go:1250]: Yes understand. I rely on Go's race detector for things setup once at creation time and never changed such that locks are not required if the underlying pointer is safe like channels.

File: server/stream.go (hunk 3 above)
Comment (derekcollison) [reply server/stream.go:1012]: I don't think so since these are channels. Just can't nil them out ever which is fine.

Comment (derekcollison) [reply server/stream.go:1012]: Read from multiple Go routines ok.

File: server/consumer.go (hunk 2 above)
Comment (kozlovic) [reply server/consumer.go:1250]: So unless consumer object is created in some tests without proper call to AddConsumer - and store is possibly not set, maybe we should remove the check to `o.store == nil` in signalNewMessages() to simply avoid confusion or second guessing.

**Review by kozlovic — APPROVED** [review]
LGTM

**Comment by derekcollison** [comment]
That is one way we tell we are closed, but could remove, won't matter. Will do that, squash and merge.

**Comment by kozlovic** [comment]
> That is one way we tell we are closed, but could remove, won't matter. Will do that, squash and merge.

Not anymore. Remember I added the `closed` boolean to avoid a race that was causing issues. But anyway it would prove my point that if we had a test that closes the consumer while there was a new message that cause signal, then we would have gotten the race detector to show the issue (if we replaced with `if o.closed {` since o.store is no longer set to nil).

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**